merlion/models/automl/autoets.py

```diff
@@ -30,3 +92,166 @@ class AutoETS(SeasonalityLayer):
     """
 
     config_class = AutoETSConfig
+
+    def generate_theta(self, train_data: TimeSeries) -> Iterator:
+        """
+        generate [theta]. theta is a list of parameter combination [error, trend, damped_trend, seasonal]
+        """
+        y = train_data.univariates[self.target_name].np_values
+
+        # check the size of y
+        n_samples = y.shape[0]
+        if n_samples <= 3:
+            self.information_criterion = "aic"
+
+        # auto-detect seasonality if desired, otherwise just get it from seasonal order
+        if self.config.auto_seasonality:
+            candidate_m = super().generate_theta(train_data=train_data)
+            self.config.seasonal_periods, _, _ = super().evaluate_theta(thetas=candidate_m, train_data=train_data)
+
+
+        # set the parameters ranges for error, trend, damped_trend and seasonal
+        if np.any(y <= 0):
+            E_range = ["add"]
+            T_range = ["add", None]
+        else:
+            E_range = ["add", "mul"]
+            if self.config.allow_multiplicative_trend:
+                T_range = ["add", "mul", None]
+            else:
+                T_range = ["add", None]
+
+        if self.config.seasonal_periods <= 1 or self.config.seasonal_periods is None or y.shape[0] <= self.config.seasonal_periods:
+            self.config.seasonal_periods = 1
+            S_range = [None]
+        elif np.any(y <= 0):
+            S_range = ["add", None]
+        else:
+            S_range = ["add", "mul", None]
+        D_range = [True, False]
+
+        if not self.config.auto_error:
+            E_range = [self.model.config.error]
+        if not self.config.auto_trend:
+            T_range = [self.model.config.trend]
+        if not self.config.auto_seasonal:
+            S_range = [self.model.config.seasonal]
+        if not self.config.auto_damped:
+            D_range = [self.model.config.damped_trend]
+
+        thetas = []
+        for error, trend, seasonal, damped in product(E_range, T_range, S_range, D_range):
+            if trend is None and damped:
+                continue
+            # if (error, trend, seasonal, damped) not in variants_pool_pred_interval:
+            #     continue
+            if self.config.additive_only:
+                if error == "mul" or trend == "mul" or seasonal == "mul":
+                    continue
+            if self.config.restrict:
+                if error == "add" and (trend == "mul" or seasonal == "mul"):
+                    continue
+                if error == "mul" and trend == "mul" and seasonal == "add":
+                    continue
+
+            thetas.append([error, trend, seasonal, damped])
```

Comment:
Based on the above comment, I think it makes most sense for each `theta` to have the form `[error, trend, seasonal, damped, seasonal_periods]`.

Reply:
I think here we follow the similar logic of `AutoSarima`, seasonal_periods are automatically detected by `autosarima_utils.multiperiodicity_detection`. Even the results have multiple periodicity, we provide a strategy to select the best one. Thus, in `generate_theta` function, we aim to explore various hyper-parameters combination based on information criterion. In AutoSarima, they are ARMA order, SARMA order (periodicity, difference order, seasonal difference order are automatically chosen before the iteration of thetas). In AutoETS, only the error, trend, seasonal require the search of candidants